sync-tags: add support for v0.x.y tags

nikhita opened the pull request:
For https://github.com/kubernetes/kubernetes/issues/84608

Implements changes to the go modules KEP made in
https://github.com/kubernetes/enhancements/pull/1350.

This commit updates sync-tags to publish `v0.x.y` tags for corresponding
`v1.x.y` Kubernetes tag. The old `kubernetes-1.x.y` tags are still also
published. Both `kubernetes-1.x.y` and `v0.x.y` point to the same sha.

For `v0.x.y`:

- At this tag, `go.mod` uses `v0.x.y` tags for dependencies.
- When this tag is manually specified in a `go.mod` file, it will be
retained as `v0.x.y`.

For `kubernetes-1.x.y`:

- At this tag, `go.mod` uses `v0.x.y` tags for dependencies. Note: this
differs from the previous way, where pseudoversions were used.
- When this tag is manually specified in a `go.mod` file, go will resolve
it to the appropriate pseudoversion.
---

Verified locally. A sample run for client-go (and api + apimachinery) is shown below:

- [v0.17.0](https://github.com/prowtest/client-go/tree/v0.17.0) = [kubernetes-1.17.0](https://github.com/prowtest/client-go/tree/kubernetes-1.17.0)
- [v0.17.0-rc.2](https://github.com/prowtest/client-go/tree/v0.17.0-rc.2) = [kubernetes-1.17.0-rc.2](https://github.com/prowtest/client-go/tree/kubernetes-1.17.0-rc.2)
- [v.17.0-beta.2](https://github.com/prowtest/client-go/tree/v0.17.0-beta.2)  = [kubernetes-1.17.0-beta.2](https://github.com/prowtest/client-go/tree/kubernetes-1.17.0-beta.2) 
- [v0.18.0-alpha.0](https://github.com/prowtest/client-go/tree/v0.18.0-alpha.0)  = [kubernetes-1.18.0-alpha.0](https://github.com/prowtest/client-go/tree/kubernetes-1.18.0-alpha.0) 


/cc @sttts @liggitt @dims 
/assign @sttts

liggitt commented on cmd/sync-tags/gomod.go:
```diff
@@ -59,10 +59,14 @@ func updateGomodWithTaggedDependencies(tag string, depsRepo []string) (bool, err
 		}
 		rev := commit.String()
-		pseudoVersion := fmt.Sprintf("v0.0.0-%s-%s", commitTime.UTC().Format("20060102150405"), rev[:12])
+		pseudoVersionOrTag := fmt.Sprintf("v0.0.0-%s-%s", commitTime.UTC().Format("20060102150405"), rev[:12])
```
as soon as we start adding tags, v0.0.0 will no longer be the correct pseudoversion for commits in between semver tags... see https://tip.golang.org/cmd/go/#hdr-Pseudo_versions

My reading of that means that when generating a non-tagged version:
* if the nearest semver tag is `v0.17.0`, we should generate a pseudoversion using the format string `v0.17.1-0.%s-%s`
* if the nearest semver tag is `v1.17.0-alpha.3`, we should generate a pseudoversion using the format string `v1.17.0-alpha.3.0.%s-%s`

sttts replied:
isn‘t there automatic normalization through go mod we can use?

liggitt replied:
that means we need to know the nearest existing tag for the current repo... not sure whether we've fetched that info and if it is available currently

liggitt replied:
> isn‘t there automatic normalization through go mod we can use?

I'm not sure how much of that is available to us before the commits/tags are pushed to the authoritative VCS

nikhita replied:
It isn't available currently, but once we fetch that looks like we can plug it into https://github.com/golang/vgo/blob/master/vendor/cmd/go/internal/modfetch/pseudo.go#L48

nikhita replied:
Updated to not use pseudoversions anymore, so this problem is fixed.

sttts replied:
but tags? why didn't we come up with that in the first place? :)

liggitt commented on cmd/sync-tags/main.go:
```diff
@@ -323,6 +306,31 @@ func main() {
 		createdTags = append(createdTags, bName)
+
+		// if publishSemverTags is true, create annotated semver tag
+		if *publishSemverTags {
```
does this script sync *all* existing tags? will this try to create semver tags corresponding to all the existing kubernetes release tags?

nikhita replied:
No, only newer tags but will make this simpler while fixing https://github.com/kubernetes/publishing-bot/pull/210#discussion_r354082521

nikhita replied:
Updated.

liggitt commented on cmd/sync-tags/main.go:
```diff
@@ -323,6 +306,31 @@ func main() {
+
+		// if publishSemverTags is true, create annotated semver tag
+		if *publishSemverTags {
+			semverTag := name[:1] + "0" + name[2:] // replace v1.x.y with v0.x.y
```
this mapping of upstream v1.x.y upstream -> v0.x.y staging is kube-specific... most of the other options in this script try to remain generic. @sttts might have thoughts on this

nikhita replied:
> this mapping of upstream v1.x.y upstream -> v0.x.y staging is kube-specific... most of the other options in this script try to remain generic. 

For the record: we discussed this on slack and decided to go with renaming the flag to `--publish-v0-semver` to represent the v1 -> v0 semver mapping.

liggitt commented on cmd/sync-tags/main.go:
```diff
@@ -323,6 +306,31 @@ func main() {
+
+			if goModChanged && len(dependentRepos) > 0 {
+				wt := checkoutBranchTagCommit(r, bh, dependentRepos)
+				changed = updateGoModAndGodeps(semverTag, dependentRepos, *generateGodeps, true)
```
by creating the prefixed tag matching upstream (kubernetes-1.x.y) first, and the semver tag second, isn't this no longer re-entrant if the script fails during the second tag creation (since we return early if bTagCommits[bName] already exists)?

liggitt replied:
also, won't calling updateGoModAndGodeps again mean that we'll change the go.mod files before adding this tag, so `kubernetes-1.x.y` and `v0.x.y` tags will not actually point to the same sha? I expected both tags to point to the same content

nikhita replied:
>  isn't this no longer re-entrant if the script fails during the second tag creation (since we return early if bTagCommits[bName] already exists)?

will fix

> also, won't calling updateGoModAndGodeps again mean that we'll change the go.mod files before adding this tag, so kubernetes-1.x.y and v0.x.y tags will not actually point to the same sha? I expected both tags to point to the same content

My understanding is that we'd want to use `v0.x.y` tags in go.mod while publishing the `v0.x.y` tags (example - https://github.com/prowtest/client-go/blob/v0.17.0-rc.2/go.mod#L41-L42).

To keep the sha same for both `kubernetes-1.x.y` and `v0.x.y` (makes sense), should we then also use `v0.x.y` tags in go.mod while publishing `kubernetes-1.x.y` i.e. same go.mod content for `kubernetes-1.17.0-rc.2` as the link above?

liggitt replied:
> To keep the sha same for both kubernetes-1.x.y and v0.x.y (makes sense), should we then also use v0.x.y tags in go.mod while publishing kubernetes-1.x.y i.e. same go.mod content for kubernetes-1.17.0-rc.2 as the link above and we won't use pseudoversions in tags anymore?

That was my expectation

nikhita replied:
Updated.

nikhita commented:
Addressed all comments except https://github.com/kubernetes/publishing-bot/pull/210#discussion_r353966692. Also, updated the main PR body and commit messages to reflect the new behaviour and added links to a new test run in the PR body.

If `--publish-semver-tags` is true, we won't use pseudoversions, so the problem described in https://github.com/kubernetes/publishing-bot/pull/210#discussion_r353916045 is not relevant anymore.

nikhita commented on cmd/sync-tags/main.go:
```diff
@@ -199,15 +206,20 @@ func main() {
-			continue
-		}
+		// if tagExists(r, bName) {
+		//	continue
```
@sttts question about this if statement: why is this required? don't we do the same check in:

https://github.com/kubernetes/publishing-bot/blob/dc65f29519695e122f8f4aa7dc1269ab6997884a/cmd/sync-tags/main.go#L202-L205

sttts replied:
wasn't the bName check above for the origin tag only? I think we rewrite tag names on fetch to `origin/<name>` or something like that. So we have local tags and origin tags. This check is for local tags, the upper for origin.

nikhita replied:
Ah right, will update.

sttts commented on cmd/sync-tags/gomod.go:
```diff
@@ -71,46 +75,46 @@ func updateGomodWithTaggedDependencies(tag string, depsRepo []string) (bool, err
 
 		tidyCommand := exec.Command("go", "mod", "tidy")
-		tidyCommand.Env = append(os.Environ(), "GO111MODULE=on", "GOPOXY=file://${GOPATH}/pkg/mod/cache/download")
+		tidyCommand.Env = append(os.Environ(), "GO111MODULE=on")
```
why is this not necessary anymore?

nikhita replied:
GOPROXY is not needed while running go mod tidy. Moreover, the original code had a typo "GOPOXY", so it was never really used. :)

liggitt replied:
> GOPROXY is not needed while running go mod tidy

really? tidy requires resolving dependencies, and this module has dependencies on unpublished tags/shas... I don't understand how this is working without GOPROXY pointing at our cache

liggitt replied:
are we inheriting a GOPROXY setting from our parent env?

nikhita replied:
> are we inheriting a GOPROXY setting from our parent env?

No, I don't think so. I tried running this locally as well and it did not globally try setting GOPROXY fwiw.

I _think_ `go mod tidy` looks at the local mod cache before donwloading from vcs...

Running the following commands at client-go checked out at `v0.18.0-alpha.0`:

- With the local mod cache cleared:

```
➜  client-go git:(2768422194b) GO111MODULE=on go mod tidy -v                                                                                                                                               0s 
get "k8s.io/apimachinery": found meta tag get.metaImport{Prefix:"k8s.io/apimachinery", VCS:"git", RepoRoot:"https://github.com/kubernetes/apimachinery"} at //k8s.io/apimachinery?go-get=1
go: k8s.io/apimachinery@v0.18.0-alpha.0: reading k8s.io/apimachinery/go.mod at revision v0.18.0-alpha.0: unknown revision v0.18.0-alpha.0
```

- After adding a `v0.18.0-alpha.0.mod` at `${GOPATH}/pkg/mod/cache/download/k8s.io/apimachinery/@v/`, `GO111MODULE=on go mod tidy -v` does not produce any output/errors.

liggitt replied:
> With the local mod cache cleared:
> ...
> go: k8s.io/apimachinery@v0.18.0-alpha.0: reading k8s.io/apimachinery/go.mod at revision v0.18.0-alpha.0: unknown revision v0.18.0-alpha.0

did that fail the tidy command? did that fail our sync-tags command? (it should)

> After adding a v0.18.0-alpha.0.mod at ${GOPATH}/pkg/mod/cache/download/k8s.io/apimachinery/@v/, GO111MODULE=on go mod tidy -v does not produce any output/errors.

are we currently creating the `v0.18.0-alpha.0.mod` file?

nikhita replied:
> did that fail the tidy command?

When I manually cleared my local cache and ran `go mod tidy`, yes.

> did that fail our sync-tags command? (it should)
> are we currently creating the v0.18.0-alpha.0.mod file?

We publish the `.mod`, `.info`, `.zip` files for each tag at the local cache, so we have a `v0.18.0-alpha.0.mod` file. And since we have this file, the sync-tags command does not fail.

https://github.com/kubernetes/publishing-bot/blob/3f3695e10c68b22812723076d843d2e48a65feaf/cmd/sync-tags/gomod.go#L80

liggitt replied:
ok, so it sounds like we're getting the implicit behavior of "check the cache dir, then hit the network", and we happen to have the things we want in the cache dir.

I still think we should formalize that by pointing GOPROXY at the cache dir (fixing the typo in the envvar) rather than removing it. If the item is not in the cache dir, I really don't want us talking to the network to try to find things.

nikhita replied:
> I still think we should formalize that by pointing GOPROXY at the cache dir (fixing the typo in the envvar, rather than removing it).

Done.

sttts commented on cmd/sync-tags/main.go:
```diff
@@ -258,71 +270,69 @@ func main() {
-			if os.IsNotExist(err) {
+			if !goModExists {
 				if _, err2 := os.Stat("Godeps/Godeps.json"); err2 == nil {
 					fmt.Printf("Updating Godeps.json to point to %s tag.\n", bName)
```
should we replace Godeps.json with go.mod everywhere?

sttts replied:
or is this really still about godeps?

nikhita replied:
This is for the case where there is no go.mod but only Godeps.json file in the repo. We don't need it for k8s repos but might be useful for non-k8s ones.

Or we could remove it here and force them to use the `godeps` branch in this repo, but that'd break compatibility.

liggitt replied:
does this script still run on 1.14.x, or did we split the job to use the godeps branch for publishing that branch? dropping godeps manipulation seems ok, but I'd do it in a separate PR

nikhita replied:
> does this script still run on 1.14.x, or did we split the job to use the godeps branch for publishing that branch?

The latter. Will remove this in another PR. :+1:

nikhita commented on cmd/sync-tags/main.go:
```diff
@@ -199,13 +206,19 @@ func main() {
+		// (we build master first, i.e. the x.y.z-alpha.0 tag on master will not be created for feature branches)
+		if *publishSemverTags && semverTagExists && nonSemverTagExists {
+			continue
+		} else if nonSemverTagExists {
```
Updated the check for tags at origin and locally. Verified this and pushed the latest results to the test repo, but can you sanity check this again?

liggitt commented on cmd/sync-tags/main.go:
```diff
@@ -488,3 +499,48 @@ func regenerateGodepsFromGoMod() error {
+	fmt.Printf("Adding extra commit fixing dependencies to point to %s tags.\n", tag)
+	publishingBotNow := publishingBot
+	publishingBotNow.When = time.Now()
+	bh, err := wt.Commit(fmt.Sprintf("Fix dependencies to point to %s tag", tag), &gogit.CommitOptions{
```
nit: prefer "Update dependencies to %s tag"

liggitt commented:
The tags published to https://github.com/prowtest look good

nikhita commented on cmd/sync-tags/main.go:
```diff
@@ -463,7 +474,7 @@ func mappingOutputFileName(fnameTpl string, branch, tag string) string {
 func regenerateGodepsFromGoMod() error {
 	goListCommand := exec.Command("go", "list", "-m", "-json", "all")
-	goListCommand.Env = append(os.Environ(), "GO111MODULE=on", "GOPOXY=file://${GOPATH}/pkg/mod/cache/download")
+	goListCommand.Env = append(os.Environ(), "GO111MODULE=on")
```
Found another GOPOXY btw :/

liggitt commented on cmd/sync-tags/main.go:
```diff
@@ -182,10 +183,16 @@ func main() {
 		if *prefix != "" {
 			bName = *prefix + name[1:] // remove the v
 		}
+		semverTag := name[:1] + "0" + name[2:] // replace v1.x.y with v0.x.y
```
at the very last, only do this swap to replace `v1.` with `v0.` (otherwise, we'd collapse v1.x.y and v2.x.y to the same tags)

```
semverTag = name
if strings.HasPrefix(name, "v1.") {
  semverTag = "v0." + strings.TrimPrefix(name, "v1.")
}
```

liggitt replied:
also, make sure name is actually a valid semver tag before triggering the semver path

nikhita replied:
Done. Also updated the results of the latest run to https://github.com/prowtest. PTAL.

liggitt commented on cmd/sync-tags/main.go:
```diff
@@ -182,11 +184,29 @@ func main() {
 	createdTags := []string{}
 	for name, kh := range kTagCommits {
+		// temporarily publish only alpha tags
+		if !strings.Contains(name, "alpha") {
```
Was this present when you ran in the prow test org? If so, how did the rc.2 tag get published?

nikhita replied:
Not present, and I manually deleted the rc2, beta2 and alpha0 tags in published repos to trigger runs for those. If not, sync-tags wouldn't sync any tags (it syncs only newer tags).

liggitt replied:
I would drop this, and just do dry runs against the test org

nikhita replied:
> I would drop this, and just do dry runs against the test org

If we drop this and merge this PR before Monday, `kubernetes-1.17.0` and `v0.17.0` would be automatically published once `v1.17.0` is tagged in k/k... I don't think that would give us enough time to do a run of `v1.17.0` on a test org.

liggitt replied:
would it make sense to:
1. separately merge a commit that skips syncing tags
2. let k8s cut v1.17.0
3. make this branch re-enable syncing tags, and do test runs against a test org
4. merge this and re-enable tag sync once ready?

that would allow us to get this in for 1.17.0 without rushing

nikhita replied:
Sounds good. :+1: 

Will make the changes in a bit.

nikhita replied:
Dropped this change and created https://github.com/kubernetes/kubernetes/pull/86008 to disable syncing tags.

liggitt commented on cmd/sync-tags/main.go:
```diff
@@ -182,11 +184,29 @@ func main() {
 
+		// check if name is a semver tag
+		// the semver lib requires the v prefix to be dropped before parsing
+		var skipPublishingSemverTag bool
```
considering both publishSemverTags and skipPublishingSemverTag in many places is confusing. suggest something more like:

```go
var (
  semverTag = ""
  publishSemverTag = false
)
// if we are publishing semver tags
if *publishSemverTags {
  // and this is a valid v1... semver tag
  if _, semverErr := semver.Parse(name[1:]); semverErr == nil && strings.HasPrefix(name, "v1.") {
    publishSemverTag = true
    semverTag = "v0." + strings.TrimPrefix(name, "v1.")
  }
}
```

and then only consider the `publishSemverTag` boolean in the following logic

nikhita replied:
Done (and tested).

liggitt commented on cmd/sync-tags/main.go:
```diff
@@ -199,13 +214,19 @@ func main() {
+			continue
+		} else if nonSemverTagExists {
 			continue
 		}
```
this bit still looks strange to me. if either tag exists at origin, we cannot make changes and should skip, right? something like this, maybe?

```go
if publishSemverTag && semverTagAtOrigin {
  continue
}
if nonSemverTagAtOrigin {
  continue
}
```

for the local checks, I'm not sure how re-entrant this script is expected to be (e.g. if it crashes after creating a local non-semver tag, but before pushing, do we wipe/reclone on next run?)

nikhita replied:
Will let @sttts confirm but this is my understanding:

> if it crashes after creating a local non-semver tag, but before pushing, do we wipe/reclone on next run?

The tag will be [appended in the push script](https://github.com/kubernetes/publishing-bot/blob/23fba6b1430ce9a03fc61bde591c4e1bdbf7eb84/cmd/sync-tags/main.go#L349) once it's created, so it'll try pushing again when the push script is executed on the next run.

>  if either tag exists at origin, we cannot make changes and should skip, right?

Because the script handles cases when tags have been created locally but not pushed, I think we'll need to consider cases when the tag exists locally as well (and not just whether it exists at origin).

> if either tag exists

I thought of skipping tag creation if _one_ of them exists but then considered the following cases...though I'm not entirely sure when we'd hit these cases. Maybe if someone deletes a tag? However, retagging on tag deletion is also a bad idea... :thinking: 

- If `kubernetes-1.x.y` exists but `v0.x.y` does not, the go.mod file for `kubernetes-1.x.y` would point to `v0.x.y` so we **need** to make sure that `v0.x.y` is published.

- If `v0.x.y` exists but `kubernetes-1.x.y` does not, we publish it because both go hand-in-hand (though not publishing it doesn't harm in any way).

sttts replied:
> The tag will be appended in the push script once it's created, so it'll try pushing again when the push script is executed on the next run.

If I remember right, the push script is recreated on each run (I hope so). Also local tags should be remove before each run.

The mixed logic for semver and non-semver irritates me. I would make all code below idem-potent. Then we could skip this early continue here, but we do it for purely performance reason to skip the git checkout magic if there is no new tag.

Then what @nikhita has in the code looks right. Could be rewritten to the following:

```
if nonSemverTagExists && (!publishSemverTag || semverTagExists) {
  continue
}
```

liggitt commented:
Overall approach and results lgtm, I just had the one comment about re-entrancy and origin tags already existing. Will defer to @sttts on that, as he is much more familiar with the flow this is executed/retried in

sttts commented on cmd/sync-tags/main.go:
```diff
@@ -258,71 +279,69 @@ func main() {
+				if publishSemverTag {
+					changed = updateGoModAndGodeps(semverTag, dependentRepos, *generateGodeps, true)
+				} else {
+					changed = updateGoModAndGodeps(bName, dependentRepos, *generateGodeps, false)
```
does this mean we don't support semver tags and `kubernetes-*` tags in parallel?

nikhita replied:
> does this mean we don't support semver tags and kubernetes-* tags in parallel?

We do. Both of them point to the same sha, so we create tags at the sha later (like you commented below).

sttts replied:
annotated tags lead to different shas. Doesn't this mean conflicting shas if I pin all dependencies to kubernetes-v1.17.0, but e.g. client-go kuberentes-v1.17.0 points to apimachinery v0.17.0 which is different from apimachiner kubernetes-v1.17.0 ?

liggitt replied:
> annotated tags lead to different shas

really? https://github.com/prowtest/client-go/releases shows identical shas for kubernetes-1.17.0-rc.2 and v0.17.0-rc.2

sttts commented on cmd/sync-tags/main.go:
```diff
@@ -258,71 +279,69 @@ func main() {
+			if err != nil {
+				glog.Fatalf("Failed to create tag %q: %v", bName, err)
+			}
+			createdTags = append(createdTags, bName)
```
all this look idem-potent and independent. What confuses me is that the commits these tags (semver and non-semver) use, are the same. I would expect different ones, pointing to the respective tags of the dependencies.

nikhita replied:
> What confuses me is that the commits these tags (semver and non-semver) use, are the same. I would expect different ones, pointing to the respective tags of the dependencies.

That was the initial approach, but that exposed us to the pseudoversion problem. Having them point to the `v0.x.y` is not ideal, but that solves the pseudoversion problem for us.

Also, having them point to different commits would mean that `kubernetes-1.x.y` and `v0.x.y` won't be equivalent (in terms of sha) anymore. Prior discussion: https://github.com/kubernetes/publishing-bot/pull/210#discussion_r354098096

liggitt replied:
> What confuses me is that the commits these tags (semver and non-semver) use, are the same. I would expect different ones, pointing to the respective tags of the dependencies.

There is little benefit to shipping a non-semver tag with a go.mod pointing to other non-semver tags, since the tags are tidied to v0.0.0-... before commit, making the go.mod files almost unusable.

nikhita commented:
Just adding an explicit hold until we've tested this thoroughly
/hold

nikhita commented:
Updated with the latest changes, PTAL. The current workflow (as discussed on slack) is:

1) figure out all the tags we want to create
2) if any already exist remotely, we're done
3) if any already exist locally, delete them (and also delete the corresponding go mod cache .mod/.zip files)
4) update+tidy+commit the go.mod (with the semver tag if we're semver-tagging, with the pseudoversion otherwise), create all tags, then push all tags with --atomic

Note: for 3) we don't overwrite the existing .mod/.info/.zip files in the go mod cache dir while building tags instead because with the current logic, we avoid creating the `.mod`, etc files if they already exist.

https://github.com/kubernetes/publishing-bot/blob/dde80ae17679962763a799410d6e54e0eb6df361/cmd/sync-tags/gomod.go#L157-L159

Even if we tried to keep track of overwriting it _once_, it isn't possible to do it across multiple runs (for multiple repos) of sync-tags.

liggitt commented on cmd/sync-tags/main.go:
```diff
@@ -199,14 +214,35 @@ func main() {
+		if tagExists(r, bName) || (publishSemverTag && tagExists(r, semverTag)) {
+			if err := deleteTag(bName); err != nil {
+				glog.Fatalf("Failed to delete tag %s: %v", bName, err)
+			}
```
won't this fail if the condition that got us into this block was `publishSemverTag && tagExists(r, semverTag)`?

liggitt replied:
I expected something more like:
```
if tagExists(r, bName) {
  if err := deleteTag(bName); err != nil {
     glog.Fatalf("Failed to delete tag %s: %v", bName, err)
  }
  if err := cleanCacheForTag(bName); err != nil {
    glog.Fatalf("Failed to clean go mod cache for %s: %v", bName, err)
  }
}
if publishSemverTag && tagExists(r, semverTag) {
  if err := deleteTag(semverTag); err != nil {
     glog.Fatalf("Failed to delete tag %s: %v", semverTag, err)
  }
  if err := cleanCacheForTag(semverTag); err != nil {
    glog.Fatalf("Failed to clean go mod cache for %s: %v", semverTag, err)
  }
}
```

liggitt replied:
it also looks like we should call cleanCacheForTag first, otherwise we won't enter this block if called again after a failure while calling cleanCacheForTag (the git tag will be gone and we won't enter this block)

nikhita replied:
Addressed these comments. PTAL.

I think there were two more things to consider:

- We can't call `cleanCacheForTag(bName)` because the cache would contain pseudoversion.mod, pseudoversion.zip files, not bName.mod, etc files. We'd need to use `cleanCacheForTag(pseudoversion)`.

- We can't call `cleanCacheForTag(pseudoversion)` if bName exists -- we'd also need to check if we are publishing semver tags. In case of semver tagging, we'll have semverTag.mod files in the cache, instead of pseudoversion.mod files.

liggitt replied:
> We can't call cleanCacheForTag(bName) because the cache would contain pseudoversion.mod, pseudoversion.zip files

Good point

> We can't call cleanCacheForTag(pseudoversion) just if bName exists -- we'd also need to check if we are publishing semver tags

cleanCacheForTag (correctly) tolerates absence of any of the files it cares about, so we should be able to call `cleanCacheForTag(pseudoversion)` unconditionally, right?

nikhita replied:
> cleanCacheForTag (correctly) tolerates absence of any of the files it cares about, so we should be able to call cleanCacheForTag(pseudoversion) unconditionally, right?

done

nikhita commented:
Rebased on top of https://github.com/kubernetes/publishing-bot/pull/211

liggitt commented:
/lgtm
/hold pending run against kubernetes-nightly org

liggitt commented:
test run to kubernetes-nightly org looks good
* all repos got v0.17.0 and kubernetes-1.17.0 tags on the same sha
* all go.mod files referenced v0.17.0 tags for peer k8s.io dependencies

liggitt commented:
/hold cancel

sttts commented:
/approve
